### PR TITLE
 FDS-219 Implementation of the function to display a list of all users and filters

### PR DESCRIPTION
**Pull request by Baraulia**

### CODE REVIEW

**Comment by TonySinero** [comment]
This is of course approve, but it will be difficult for mentors to go through, primarily because of the large volume of changes. But the worst thing is that there are a lot of commits with the same name and therefore sometimes difficult to track.

**Review by Nalivayko13 — COMMENTED** [review]
LGTM

File: GRPC/auth.proto [hunk 1]
```diff
@@ -9,6 +9,7 @@ service Auth {
   rpc BindUserAndRole(User) returns (ResultBinding) {}
   rpc TokenGenerationByRefresh(RefreshToken) returns (GeneratedTokens) {}
   rpc TokenGenerationByUserId(User) returns (GeneratedTokens) {}
+  rpc GetAllRoles(Request) returns (Roles) {}
```
Comment (Ann40a) [inline GRPC/auth.proto]: Can this be replaced with `google.protobuf.Empty`?

File: handler/router.go [hunk 2]
```diff
@@ -31,8 +31,8 @@ func (h *Handler) InitRoutes() *gin.Engine {
 	{
 		user.GET("/:id", h.getUser)
 		user.DELETE("/:id", h.deleteUserByID)
-		user.PUT("/:id", h.updateUser)
-		user.GET("/", h.getUsers)
+		user.PUT("/", h.updateUser)
+		user.POST("/", h.getUsers)
```
Comment (Ann40a) [inline handler/router.go]: Why it became post? Almost all filters can be passed in query parameters. There are special approaches for passing multiple values (arrays) when needed. Just search for a library. I did quick search and found one, but not very popular (https://github.com/pasztorpisti/qs)
image

File: model/filters_model.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	}
+	return []byte(fmt.Sprintf(`"%s"`, c.Time.Format(Layout))), nil
+}
+
+type RequestFilters struct {
+	ShowDeleted bool   `json:"show_deleted"`
+	FilterData  bool   `json:"filter_data"`
+	StartTime   MyTime `json:"start_time"`
+	EndTime     MyTime `json:"end_time"`
+	FilterRole  string `json:"filter_role"`
```
Comment (Ann40a) [inline model/filters_model.go]: Just `role` should be fine. When you see some field in get request you understand that it is a filter by this field.

**Comment by TimofeiShorokhov** [comment]
approve

File: pkg/database/db.go [hunk 4]
```diff
@@ -42,6 +42,8 @@ const USER_SCHEMA = `
 		id serial not null primary key ,
 		email varchar(225) NOT NULL UNIQUE,
 		password varchar(225) NOT NULL,
-	  	created_at timestamp NOT NULL
-	);
+	    role varchar(50) NOT NULL,
+	  	created_at date NOT NULL,
+	    deleted bool NOT NULL
```
Comment (Ann40a) [inline pkg/database/db.go:47]: please fix formatting

File: repository/user_repository.go [hunk 5]
```diff
@@ -53,54 +52,191 @@ func (u *UserPostgres) GetUserAll(page int, limit int) ([]model.ResponseUser, in
...
+	if page == 0 || limit == 0 {
+		if filters.ShowDeleted {
+			query := "SELECT id, email, role, created_at FROM users WHERE role = $1 ORDER BY id"
+			rows, err = transaction.Query(query, filters.FilterRole)
+			if err != nil {
+				u.logger.Errorf("GetUserAll: can not executes a query:%s", err)
+				return nil, 0, fmt.Errorf("getUserAll:repository error:%w", err)
+			}
+		} else {
+			query := "SELECT id, email, role, created_at FROM users WHERE deleted = false AND role = $1"
```
Comment (Ann40a) [inline repository/user_repository.go]: it also needs ORDER BY

**Review by Ann40a — APPROVED** [review]

File: GRPC/auth.proto (hunk 1 above)
Comment (Baraulia) [reply GRPC/auth.proto]: Yes, it can

Comment (Baraulia) [reply GRPC/auth.proto]: Fixed

File: model/filters_model.go (hunk 3 above)
Comment (Baraulia) [reply model/filters_model.go]: Fix it

File: pkg/database/db.go (hunk 4 above)
Comment (Baraulia) [reply pkg/database/db.go:47]: fix it

File: repository/user_repository.go (hunk 5 above)
Comment (Baraulia) [reply repository/user_repository.go]: fixed

File: handler/router.go (hunk 2 above)
Comment (Baraulia) [reply handler/router.go]: fixed

**Pull request merged**